File: Libreria.java
import java.util.ArrayList;

public class Libreria {
    ArrayList<Libro> books = new ArrayList<>();
    ArrayList<Usuario> users = new ArrayList<>();

    public void addUser(Usuario user){
        users.add(user);
    }

    public void addBook(Libro book){
        books.add(book);
    }

    public void rent(int userID, int bookID){
        boolean userFound = false;
        for (Usuario user : users){
            if (user.getId() == userID){
                userFound = true;
                boolean bookFound = false;
                for (Libro book : books){
                    if (book.getId() == bookID){
                        bookFound = true;
                        if (!book.isRented()) {
                            book.setRented(true);
                            user.rentBook(book);
                            System.out.println("Libro rentado con éxito.");
                        }
                        else{
                            System.out.println("Este libro ya está siendo rentado.");
                        }
                    }
                }
                if(!bookFound){
                    System.out.println("Libro no encontrado.");
                }
            }
        }
        if(!userFound){
            System.out.println("Usuario no encontrado.");
        }
    }

    public void listUsers(){
        System.out.println("============= USUARIOS REGISTRADOS =============");
        if (users.isEmpty()){
            System.out.println("================================================");
            System.out.println("No hay ningún usuario registrado");
            System.out.println("================================================");
        }
        else {
            for (Usuario user : users) {
                System.out.println("================================================");
                System.out.printf("ID del usuario: %s%n", user.getId());
                System.out.printf("Nombre de usuario: %s %s%n", user.getName(), user.getLastName());
                System.out.println("================================================");
            }
        }
    }

    public void listBooks(){
        System.out.println("============== LIBROS REGISTRADOS ==============");
        if (books.isEmpty()){
            System.out.println("================================================");
            System.out.println("No hay ningún libro registrado");
            System.out.println("================================================");
        }
        else {
            for (Libro book : books) {
                System.out.println("================================================");
                System.out.printf("ID del libro: %s%n", book.getId());
                System.out.printf("Titulo del libro: %s%n", book.getTitle());
                System.out.printf("Autor del libro: %s%n", book.getAuthor());
                System.out.println("================================================");
            }
        }
    }

    public void listUsersWithBooks(){
        System.out.println("=========== USUARIOS RENTANDO LIBROS ===========");
        boolean rentando = false;
        for(Usuario user : users){
            if (user.hasRentedBooks()) {
                rentando = true;
                System.out.println("================================================");
                System.out.printf("ID del usuario: %s%n", user.getId());
                System.out.printf("Nombre de usuario: %s %s%n", user.getName(), user.getLastName());
                System.out.println("Cantidad de libros rentados: " + user.booksQuantity());
                System.out.println("================================================");
            }
        }
        if (!rentando){
            System.out.println("================================================");
            System.out.println("Ningún usuario está rentando libros");
            System.out.println("================================================");
        }
    }

    public void rentedBooks(){
        System.out.println("================ LIBROS RENTADOS ================");
        boolean rentado = false;
        for(Libro book : books){
            if (book.isRented()) {
                rentado = true;
                System.out.println("=================================================");
                System.out.printf("ID del libro: %s%n", book.getId());
                System.out.printf("Titulo del libro: %s%n", book.getTitle());
                System.out.printf("Autor del libro: %s%n", book.getAuthor());
                System.out.println("=================================================");
            }
        }
        if (!rentado){
            System.out.println("=================================================");
            System.out.println("Ningún libro ha sido rentado");
            System.out.println("=================================================");
        }
    }

    public void notRentedBooks(){
        System.out.println("============== LIBROS NO RENTADOS ==============");
        boolean rentado = true;
        for(Libro book : books){
            if (!book.isRented()) {
                rentado = false;
                System.out.println("================================================");
                System.out.printf("ID del libro: %s%n", book.getId());
                System.out.printf("Titulo del libro: %s%n", book.getTitle());
                System.out.printf("Autor del libro: %s%n", book.getAuthor());
                System.out.println("================================================");
            }
        }
        if (rentado){
            System.out.println("================================================");
            System.out.println("Todos los libros han sido rentados");
            System.out.println("================================================");
        }
    }
}

File: libreria.py
class Libreria:
    __users = []
    __books = []

    def add_user(self, user):
        self.__users.append(user)

    def add_book(self, book):
        self.__books.append(book)

    def rent(self, user_id, book_id):
        user_found = False
        for user in self.__users:
            if user.get_id() == user_id:
                user_found = True
                book_found = False
                for book in self.__books:
                    if book.get_id() == book_id:
                        book_found = True
                        if not book.is_rented():
                            book.set_rented(True)
                            user.rent_book(book)
                            print("Libro rentado con éxito.")
                        else:
                            print("Este libro ya está siendo rentado.")
                if not book_found:
                    print("Libro no encontrado.")
        if not user_found:
            print("Usuario no encontrado.")

    def list_users(self):
        print("============= USUARIOS REGISTRADOS =============")
        if len(self.__users) == 0:
            print("================================================")
            print("No hay ningún usuario registrado")
            print("================================================")
        else:
            for user in self.__users:
                print("================================================")
                print("ID del usuario: {}".format(user.get_id()))
                print("Nombre de usuario: {} {}".format(user.get_name(), user.get_last_name()))
                print("================================================")

    def list_books(self):
        print("============== LIBROS REGISTRADOS ==============")
        if len(self.__books) == 0:
            print("================================================")
            print("No hay ningún libro registrado")
            print("================================================")
        else:
            for book in self.__books:
                print("================================================")
                print("ID del libro: {}".format(book.get_id()))
                print("Título del libro: {}".format(book.get_title()))
                print("Autor del libro: {}".format(book.get_author()))
                print("================================================")

    def list_users_with_books(self):
        print("=========== USUARIOS RENTANDO LIBROS ===========")
        rentado = False
        for user in self.__users:
            if user.has_rented_books():
                rentado = True
                print("================================================")
                print("ID del usuario: {}".format(user.get_id()))
                print("Nombre del usuario: {} {}".format(user.get_name(), user.get_last_name()))
                print("Cantidad de libros rentados: {}".format(user.books_quantity()))
                print("================================================")
        if not rentado:
            print("================================================")
            print("Ningún usuario está rentando libros")
            print("================================================")

    def rented_books(self):
        print("================ LIBROS RENTADOS ================")
        rentado = False
        for book in self.__books:
            if book.is_rented:
                rentado = True
                print("=================================================")
                print("ID del libro: {}".format(book.get_id()))
                print("Título del libro: {}".format(book.get_title()))
                print("Autor del libro: {}".format(book.get_author()))
                print("=================================================")
        if not rentado:
            print("=================================================")
            print("Ningún libro ha sido rentado")
            print("=================================================")

    def not_rented_books(self):
        print("============== LIBROS NO RENTADOS ==============")
        rentado = True
        for book in self.__books:
            if not book.is_rented:
                rentado = False
                print("================================================")
                print("ID del libro: {}".format(book.get_id()))
                print("Título del libro: {}".format(book.get_title()))
                print("Autor del libro: {}".format(book.get_author()))
                print("================================================")
        if rentado:
            print("================================================")
            print("Todos los libros han sido rentados")
            print("================================================")
